src/plugins/cpptemplate/cpptemplate.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ *
+ * @brief Source for cpptemplate plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+
+#include "cpptemplate.hpp"
+#include "cpptemplate_delegate.hpp"
+
+#include <kdbhelper.h>
+
+using elektra::CppTemplateDelegate;
+
+using CppKey = kdb::Key;
+using CppKeySet = kdb::KeySet;
+
+namespace
+{
+
+/**
+ * @brief This function returns a key set containing the plugin contract.
+ *
+ * @return A key set specifying the capabilities of the plugin
+ */
+CppKeySet getContract ()
+{
+	return CppKeySet{ 30,
+			  keyNew ("system/elektra/modules/cpptemplate", KEY_VALUE, "cpptemplate plugin waits for your orders", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports", KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/open", KEY_FUNC, elektraCppTemplateOpen, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/close", KEY_FUNC, elektraCppTemplateClose, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/get", KEY_FUNC, elektraCppTemplateGet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/set", KEY_FUNC, elektraCppTemplateSet, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/error", KEY_FUNC, elektraCppTemplateError, KEY_END),
+			  keyNew ("system/elektra/modules/cpptemplate/exports/checkconf", KEY_FUNC, elektraCppTemplateCheckConfig, KEY_END),
+#include ELEKTRA_README (cpptemplate)
+			  keyNew ("system/elektra/modules/cpptemplate/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END),
+			  KS_END };
+}
+
+} // end namespace
+
+extern "C" {
+
+typedef Delegator<CppTemplateDelegate> delegator;
+
+/** @see elektraDocOpen */
+int elektraCppTemplateOpen (Plugin * handle, Key * key)
+{
+	// After the call to `delegator::open` you can retrieve a pointer to the delegate via `coderDelegator::get (handle)`
+	return delegator::open (handle, key);
+}
+
+/** @see elektraDocClose */
+int elektraCppTemplateClose (Plugin * handle, Key * key)
+{
+	return delegator::close (handle, key);
+}
+
+/** @see elektraDocGet */
+int elektraCppTemplateGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * parentKey)
+{
+	CppKeySet keys{ returned };
+	CppKey parent{ parentKey };
+	bool updated = false;
+
+	if (parent.getName () == "system/elektra/modules/cpptemplate")
+	{
+		keys.append (getContract ());
+		updated = true;
+	}
+
+	parent.release ();
+	keys.release ();
+	return updated ? ELEKTRA_PLUGIN_STATUS_SUCCESS : ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
+}
+
+/** @see elektraDocSet */
+int elektraCppTemplateSet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UNUSED, Key * parentKey ELEKTRA_UNUSED)
+{
+	return ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
```

Comment:
No delegator?

Reply:
Since this is only a template I think this makes sense. The plugin writer can add a call to whatever function of the delegate she or he likes later.

Reply:
For plugin templates it makes sense that all functions are present and users remove the functions they do not need.

Reply:
In the latest commit I added an example method to the delegate. Since the design of the delegate is not really dependent on the plugin interface, I do not think it makes sense to add a method for each part of the plugin interface. If you think differently, then you are of course welcome to extend the template accordingly.

Reply:
Thank you!

To avoid code duplication from the code generator is more important anyway. Ideally we would also get rid of the manual export of symbols. With supermacros it should be possible to generate the code as needed (for both contract and symbol exporting). It could look like:

```
#define PLUGIN_HAS_OPEN
#define PLUGIN_HAS_GET
#define PLUGIN_HAS_SET

#define PLUGIN_MACRO EXPORT_CONTRACT
#include <kdbplugingen.hpp>

#define PLUGIN_MACRO EXPORT_FUNCTIONS
#include <kdbplugingen.hpp>

#define PLUGIN_MACRO EXPORT_SYMBOLS
#include <kdbplugingen.hpp>
```

By default it generates everything, so the one line `#include <kdbplugingen.hpp>` would be enough.

For more info about supermacros, read: http://s11n.net/papers/supermacros_cpp.html

Similar techniques are not rare, e.g., search for BOOST_SERIALIZATION_NVP

Reply:
Thank you for the information, but I think I will leave this task up to you (or anybody else who likes to work on this stuff).